proposals/0672-tuist-react-native.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Tuist.io in React Native
+author:
+- Riccardo Cipolleschi
+date: 2023-06-19
+---
+
+# RFC0672: Tuist.io in React Native 
+
+## Summary
+
+Managing Xcodeproj files is error prone and time consuming. 
+
+The current proposal suggests to move away from manually managing `Xcodeproj` files in favor of a declarative description leveraging [Tuist.io](https://tuist.io) as tool.
+
+## Basic example
+
+You can have a look at it in action in this PR:
+
+- https://github.com/facebook/react-native/pull/37952
+
+The idea is to describe the project with a declarative approach using Swift. Tuist allows you to load the `Project.swift` file in an instance of Xcode, leveraging also code completion and type checking to ensure the correctness of the project.
+
+## Motivation
+
+Managing Xcodeproj files is error prone and time consuming. Those files are source of frustrations for teams using them and they have been reported as one of the main pain point when users have to migrate to a new version of React Native.
+
+They are also an old way to manage projects. Even Apple realized that they are not the best approach and started developing Swift Package Manager (SwiftPM) that replace the project for developing libraries.
+
+Moder approaches relies on declarative syntax to describe how the project should look like and tools that can generate a valid Xcodeproj starting from that description.
+
+There are several tools that we can use, but the proposed one is [Tuist.io](https://tuist.io).
+
+[Tuist.io](https://tuist.io) is OSS and currently maintained by several companies, among which Shopify, which is already a React Native partner.
+
+It allows the users to describe their Apple projects in Swift (the default language for iOS), and it also leverage Xcode for type checking and code completion, ensuring that the project definition is valid at compile time.
+
+## Detailed design
+
+To adopt Tuist in React Native, we need to follow some steps:
+
+1. Land https://github.com/facebook/react-native/pull/37952. This PR already contains a functioning project which passed all the iOS tests in CI.
+2. Update the CLI to install Tuist in the user machine. Tuist comes directly with a `tuistenv`additional tool, which handle the environment and version for the user.
+3. Update the CLI and instruct it to swap the `HelloWorld` strings in the `template/ios/Package.swift` to  the AppName chosen by the user.
+4. Update the CLI to run `tuist generate` to create the project
+5. Update doctor to check that the user have tuist installed
+6. Remove the old Xcodeproj file and set the right `.gitignore` settings.
```

Comment:
Once this lands, I'd like to explore removing the dependency with CocoaPods and Ruby. Developers would only need to have Node and a package manager like NPM in their environments. Solving that problem would require:

- Including a `Project.swift` in every React Native dependency.
- Adjusting the Tuist graph resolution to look up packages under `node_modules`.

We can even go further and support defining the project in the `package.json`:

```json
{
  "tuist": {
    "project": {
       targets: []
     }
  }
}
```

Reply:
That's not possible at the moment: we use cocoapods to install some local dependencies shipped by React Native, to run some Codegen and to configure the workspace for the user.
All the libraries in the ecosystem are also leveraging Cocoapods unfortunately.

Not that we won't be able to do it ever, but:
1. It will require a lot of work on everyone's side
2. It will require and ecosystem migration (and we are already going through one right now)
3. I think that to remove Ruby and Cocoapods, the best alternative would be Swift Package Manager.